packages/firestore/src/remote/backoff.ts

```diff
@@ -92,18 +93,34 @@ export class ExponentialBackoff {
 
     // First schedule using the current base (which may be 0 and should be
     // honored as such).
-    const delayWithJitterMs = this.currentBaseMs + this.jitterDelayMs();
+    const desiredDelayWithJitterMs = Math.floor(
+      this.currentBaseMs + this.jitterDelayMs()
+    );
+
+    // Some time may have already elapsed so account for that.
+    const delaySoFarMs = new Date().getTime() - this.lastAttempt.getTime();
```

Comment:
This makes me nervous: `new Date().getTime()` is not guaranteed to be a monotonic time source. At least it's guaranteed to be in UTC so there's no timezone bug here, but large time adjustments backwards could result in negative values for `delaySoFarMs`. We should probably ensure this is no smaller than zero so the next line doesn't end up with arbitrarily large delays.

If this class accepted a ticker constructor parameter that returned a new timestamp in ms on each call this logic could be tested.

Reply:
Fair point.  Reworked to ensure it's >= 0.  I don't love the way this code looks (doing `Math.max(0, ...)` twice in a row for different reasons) but I tried rewriting it a couple different ways and it didn't get much better, so I'm just relying on comments.

I _almost_ considered doing the ticker thing and adding tests, but I think we'd need a mockable random number generator too and if I do that I won't get this ported to the other 2 clients before going OOO.  So I just tested it manually again. 😛

If you think it's time to bite the bullet and automate tests, I can though.

Reply:
I'm OK this as is.